### PR TITLE
fix: reject negative probability values in is_valid_cpd

### LINKED ISSUE #3050: fix: check_model accepts negative probability values in CPDs
check_model returns True for CPDs with negative values like [-0.5, 1.5] because it only checks that columns sum to 1. Probabilities must be non-negative.

### PR DESCRIPTION
**Pull request by PewterZz**
Fixes #3050

# The following checklist is mandatory.

Your PR will be closed if you remove the checklist or do not answer the
questions to a satisfactory level. Use of LLM is **strictly forbidden**
for any part of this checklist (even for improving language).

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing
Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its
defined scope? If you are still working on the PR, mark it as draft.
- [ ] Are all the GitHub Actions checks passing? If not, mark your PR as
draft while you fix it.

Please answer the following questions:

- Did you use an LLM for any assistance? Please describe how and what
you used it for?

Yes, I have used AI to read through the codebase and understand its structure as well as suggest best practice code implementations for the specific issue. All changes made have been tested and reviewed by me.

- Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

Yes. is_valid_cpd in DiscreteFactor.py now checks that all values are >= 0 before checking they sum to 1. Second, check_model in DiscreteBayesianNetwork.py adds an explicit check for negative values with a clear error message before calling is_valid_cpd. This way users get "contains negative probability values" instead of the generic "not equal to 1" message. Its not a common issue for most but would reduce confusion.

- What steps have you taken to verify that the changes correctly address
the issue? And what edge cases have you considered?

ran pytest pgmpy/tests/test_models/test_DiscreteBayesianNetwork.py::TestBayesianNetworkCPD::test_check_model_negative_cpd -v . it passes. tested manually that [-0.5, 1.5] raises ValueError, that [0.4, 0.6] still passes, that [0.0, 1.0] still passes, that the existing test_check_model tests still pass.

- Has the LLM added try-except blocks? They will need to be removed; any
error handling must be explicit.

None 

- Have you used LLM for generating tests? They need to be compressed
into a smaller number of tests without reducing coverage.

Partially, only 1 test

### Issue number(s) that this pull request fixes
- Fixes https://github.com/pgmpy/pgmpy/issues/3050
- Partially addresses https://github.com/pgmpy/pgmpy/issues/2202

### List of changes to the codebase in this pull request
1. Added non-negativity check in pgmpy/factors/discrete/DiscreteFactor.py (is_valid_cpd returns False for negative values)
2. Added explicit error message for negative CPD values in pgmpy/models/DiscreteBayesianNetwork.py (check_model raises "CPD for {node} contains negative probability values")
3. Added test test_check_model_negative_cpd in pgmpy/tests/test_models/test_DiscreteBayesianNetwork.py

### CODE REVIEW

**Comment by DARHWOLF** [comment]
May I request you to please put back the PR checklist. It is to be filled out and clearly mentions that it is not to be removed.

**Comment by PewterZz** [comment]
Apologies, it has been re-edited back

File: pgmpy/models/DiscreteBayesianNetwork.py
Comment (DARHWOLF) [inline pgmpy/models/DiscreteBayesianNetwork.py:464]: Could you add in the documentation the new check that is being implemented

**Review by DARHWOLF — COMMENTED** [review]
Thanks for the PR
- it might be a good idea to add a small test for `DiscreteFactor` as well. 
- it would be nice to document the new test being implemented. 

I wonder why codecov didn't fail for the missing coverage when tests were not implemented in the previous commits.

File: pgmpy/factors/discrete/DiscreteFactor.py [hunk 1]
```diff
@@ -931,6 +931,8 @@ def is_valid_cpd(self):
         """
         Checks if the factor's values can be used for a valid CPD.
         """
+        if not config.get_compute_backend().all(self.values >= 0):
```
Comment (ankurankan) [inline pgmpy/factors/discrete/DiscreteFactor.py]: Doesn't simply doing `(self.values >= 0).all()` work here? If not, add a new method to `compat_fns.py`.

File: pgmpy/factors/discrete/DiscreteFactor.py [hunk 2]
```diff
@@ -931,6 +931,8 @@ def is_valid_cpd(self):
...
+            return False
         return config.get_compute_backend().allclose(
```
Comment (ankurankan) [inline pgmpy/factors/discrete/DiscreteFactor.py]: Can you change this to use `compat_fns.allclose`.

File: pgmpy/models/DiscreteBayesianNetwork.py [hunk 3]
```diff
@@ -486,6 +486,10 @@ def check_model(self) -> bool:
                 if len(set(cpd.variables) - set(cpd.state_names.keys())) > 0:
                     raise ValueError(f"CPD for {node} doesn't have state names defined for all the variables.")
 
+                # Check if the CPD contains negative values.
+                if (cpd.get_values() < 0).any():
+                    raise ValueError(f"CPD for {node} contains negative probability values.")
+
```
Comment (ankurankan) [inline pgmpy/models/DiscreteBayesianNetwork.py]: This is not a property of the model and shouldn't be checked like this. Since, it is a property of the CPD, it should be cheked withing `is_valid_cpd` call in the line below.

**Review by ankurankan — CHANGES_REQUESTED** [review]
@PewterZz I don't think the check should be in any of these methods. Rather, the most suitable place is in `TabularCPD.__init__` method. It should just not allow the creation of a CPD object with negative values. The init method already does a bunch of checks on the values; this would be another check.

**Comment by PewterZz** [comment]
@ankurankan Updated. Moved check to `TabularCPD.__init__`, removed from `check_model` and `is_valid_cpd`, switched to `compat_fns.allclose`. Should `is_valid_cpd` also guard against negatives?

File: pgmpy/factors/discrete/CPD.py [hunk 4]
```diff
@@ -47,6 +47,11 @@ class TabularCPD(DiscreteFactor):
         match the order in the values array. If state_names is not specified,
         auto-assigns state names starting from 0.
 
+    Raises
+    ------
+    ValueError
+        If any value in ``values`` is negative.
+
```
Comment (ankurankan) [inline pgmpy/factors/discrete/CPD.py]: Either document all the errors or remove this.

File: pgmpy/tests/test_models/test_DiscreteBayesianNetwork.py [hunk 5]
```diff
@@ -731,6 +731,11 @@ def test_check_model(self):
 
         self.assertTrue(self.G.check_model())
 
+    def test_negative_cpd_rejected_at_init(self):
+        self.assertRaises(
+            ValueError, TabularCPD, "A", 2, [[-0.5], [1.5]]
+        )
+
```
Comment (ankurankan) [inline pgmpy/tests/test_models/test_DiscreteBayesianNetwork.py]: Again, this is not a property of DiscreteBN so shouldn't be tested in DiscreteBN tests.

File: pgmpy/tests/test_factors/test_discrete/test_Factor.py [hunk 6]
```diff
@@ -1110,6 +1110,15 @@ def test_cpd_init_value_not_2d(self):
...
 
+    def test_cpd_init_negative_values(self):
+        self.assertRaises(
+            ValueError,
+            TabularCPD,
+            "A",
+            2,
+            [[-0.5], [1.5]],
+        )
+
```
Comment (ankurankan) [inline pgmpy/tests/test_factors/test_discrete/test_Factor.py:1156]: This test should be in `TestTabularCPDInit` class along with the other exception tests.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by ankurankan** [comment]
Closing due to inactivity. Please feel free to reopen it if you are still working on it.

**Comment by PewterZz** [comment]
Hi @ankurankan, I pushed a follow up update on this branch and would still like this contribution reviewed. Could you please reopen this PR?

**Comment by ankurankan** [comment]
@PewterZz The tests are still failing for this PR. Can you investigate what is happening and fix it if possible? Converting to draft till all checks pass.

**Comment by PewterZz** [comment]
Hi @ankurankan, thanks for flagging this. I found the issue. The non negative guard added in this PR was also rejecting NaN values, which broke the improper prior estimator tests. I updated the check to reject only true negative values and keep NaN behavior unchanged for improper priors. I reran the affected suites locally, including the failing test_estimate_cpd_improper_prior tests, and they now pass. Could you please take another look?

**Comment by PewterZz** [comment]
Hi @ankurankan, I've addressed your review feedback (moved the negative value check to `TabularCPD.__init__` as you suggested) and all CI checks are passing. Could you take another look when you have a chance?

**Review by ankurankan — APPROVED** [review]
@PewterZz Thanks for the PR. Looks good to me.
